hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java

```diff
@@ -1670,20 +1674,71 @@ public static Tuple nftTransfer(
 
     public static List<HapiSpecOperation> convertHapiCallsToEthereumCalls(
-            final List<HapiSpecOperation> ops, final String privateKeyRef, final Key adminKey, final long defaultGas) {
+            final List<HapiSpecOperation> ops,
+            final String privateKeyRef,
+            final Key adminKey,
+            final long defaultGas,
+            final HapiSpec spec) {
         final var convertedOps = new ArrayList<HapiSpecOperation>(ops.size());
         for (final var op : ops) {
-            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall()) {
+            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall(spec)) {
+                // if we have function params, try to swap the long zero address with the EVM address
+                if (callOp.getParams().length > 0 && callOp.getAbi() != null) {
+                    var convertedParams = tryToSwapLongZeroToEVMAddresses(callOp.getParams(), spec);
+                    callOp.setParams(convertedParams);
+                }
                 convertedOps.add(new HapiEthereumCall(callOp));
+
             } else if (op instanceof HapiContractCreate callOp && callOp.isConvertableToEthCreate()) {
-                convertedOps.add(new HapiEthereumContractCreate(callOp, privateKeyRef, adminKey, defaultGas));
+                // if we have constructor args, update the bytecode file with one containing the args
+                if (callOp.getArgs().isPresent() && callOp.getAbi().isPresent()) {
+                    var convertedArgs =
+                            tryToSwapLongZeroToEVMAddresses(callOp.getArgs().get(), spec);
+                    callOp.setArgs(Optional.of(convertedArgs));
+                    convertedOps.add(updateInitCodeWithConstructorArgs(
+                            callOp.getContract(),
+                            callOp.getAbi().get(),
+                            callOp.getArgs().get()));
+                }
+
+                var createEthereum = withOpContext((spec1, logger) -> {
+                    var createTxn = new HapiEthereumContractCreate(callOp, privateKeyRef, adminKey, defaultGas);
+                    allRunFor(spec1, createTxn);
+                    // if create was successful, save the EVM address to the registry, so we can use it in future calls
+                    if (spec1.registry().hasContractId(callOp.getContract())) {
+                        allRunFor(
+                                spec1,
+                                getContractInfo(callOp.getContract()).saveEVMAddressToRegistry(callOp.getContract()));
+                    }
+                });
+                convertedOps.add(createEthereum);
+
             } else {
                 convertedOps.add(op);
             }
         }
         return convertedOps;
     }
 
+    private static Object[] tryToSwapLongZeroToEVMAddresses(Object[] args, HapiSpec spec) {
+        return Arrays.stream(args)
+                .map(arg -> {
```

Comment:
I like using a stream + `map` here but jeez it's kind of ugly.  Can't think of any easy way to clean it up either.  Sigh.

Maybe it would be an improvement to pull lines L1727-L1735 to a separate method, leaving only the test `instanceof Address` inside the `map` ...   ???

Reply:
Changed 👍